proposals/0000-overloaded-record-fields.rst

```diff
@@ -0,0 +1,352 @@
+                 implements the described feature.
+
+Overloaded Record Fields
+========================
+
+This is a proposal to introduce a new extension,
+``OverloadedRecordFields``, to make it easier to work with record
+datatypes that reuse the same field names.  It also makes some changes
+to the existing ``OverloadedLabels`` extension for consistency.
+
+
+Motivation
+----------
+
+A serious limitation of the Haskell record system is the inability to
+overload field names in record types: for example, if the data types
+
+::
+
+  data Person  = Person  { personId :: Int, name :: String }
+  data Address = Address { personId :: Int, address :: String }
+
+are in scope in the same module, there is no way to determine which
+type an occurrence of the ``personId`` record selector refers to.  A
+common workaround is to use a unique prefix for each record type, but
+this leads to less clear code and obfuscates relationships between
+fields of different records.  Qualified names can be used to
+distinguish record selectors from different modules, but using one
+module per record is often impractical.
+
+DuplicateRecordFields
+~~~~~~~~~~~~~~~~~~~~~
+
+The ``DuplicateRecordFields`` extension, introduced in GHC 8.0.1,
+makes it possible for the same field to be declared multiple times in
+a single module.  (Without it, ``Person`` and ``Address`` cannot even
+be defined simultaneously.)  However, its ability to disambiguate
+record selectors is deliberately limited to information that is
+immediately apparent during type *checking*, in the interests of
+predictable type inference behaviour.
+
+For example, the following are accepted when both occurrences of
+``personId`` are in scope:
+
+::
+
+  \ p -> personId (p :: Person)         -- explicit type signature on argument
+  \ p -> (personId :: Person -> Int) p  -- type pushed in to selector
+
+  f :: Person -> Int
+  f = personId                          -- another way to push type in to selector
+
+However, the following are rejected:
+
+::
+
+  \ (p :: Person) -> personId p  -- type of argument not used, even if obvious
+  personId (Person 1 "Me")       -- again, argument is not an explicit signature
+
+See `Trac #11343 <https://ghc.haskell.org/trac/ghc/ticket/11343>`_ for
+discussion of this aspect of ``DuplicateRecordFields``.
+
+We need some way to write polymorphic record projections, so that
+the ambiguous selector is resolved using type *inference*.
+
+From OverloadedLabels to OverloadedRecordFields
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``OverloadedLabels`` extension, also introduced in GHC 8.0.1,
+provides a way to indicate that an identifier should be resolved using
+type inference.  This extension provides a new syntax (using a ``#``
+prefix) for "labels", identifiers whose meaning depends on their type
+as well as their name.
+
+The current proposal makes it possible to interpret a label
+(e.g. ``#personId``) as a record selector, determining the record and
```

Comment:
What would be the inferred type of the single token `#personId`?  Am I right in guessing that it would be?

```haskell
#person :: HasField "personId" r a => r -> a
```